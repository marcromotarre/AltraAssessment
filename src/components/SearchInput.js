import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import SearchSVG from '../svg/search.svg';

export const SearchContainer = styled.div`
  background-color: #ffffff;
  border-bottom: 1px solid #e6e6e6;
  height: 40px;
  width: 100%;
  display: grid;
  grid-template-columns: 40px auto;
  grid-template-rows: 40px;
  column-gap: 10px;
`;

const SearchIcon = styled(SearchSVG)`
  justify-self: center;
  align-self: center;
`;

const SearchInputSearcher = styled.input`
  justify-self: start;
  align-self: center;
  font-family: 'HelveticaNeue-Thin';
  color: #929292;
  font-size: 12px;
  width: 90%;
  height: 90%;
  :focus {
    outline: none;
  }
`;

export function useInputValue(onSearch, initValue = '') {
  const [inputValue, setInputValue] = useState(initValue);
  const handleInputChange = (event) => {
    setInputValue(event.target.value);
    onSearch(event.target.value);
  };
  return { inputValue, handleInputChange };
}

const SearchInput = ({ onSearch, initValue }) => {
  const { inputValue, handleInputChange } = useInputValue(onSearch, initValue);

  return (
    <SearchContainer className>
      <SearchIcon width={21.42} data-testid="search-icon" />
      <SearchInputSearcher
        value={inputValue}
        onChange={handleInputChange}
        data-testid="search-input"
      />
    </SearchContainer>
  );
};

SearchInput.propTypes = {
  onSearch: PropTypes.func,
  initValue: PropTypes.string,
};

SearchInput.defaultProps = {
  onSearch: () => {},
  initValue: '',
};

export default SearchInput;
