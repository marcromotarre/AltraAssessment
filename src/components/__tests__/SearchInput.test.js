import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import SearchInput, { useInputValue } from '../SearchInput';

/** this component should:
 *  - allow user to write characters in input
 *  - call a search function every time that users write a character
 */

//jest mock es6 import
// see svg parcel library

//https://diessi.ca/blog/how-to-exclude-css-images-anything-from-unit-tests/
describe('Search Component tests', () => {
  const props = {
    onSearch: jest.fn(),
  };

  const event = {
    target: {
      value: 'text',
    },
  };

  it('render SearchInput Snapshot', () => {
    const tree = renderer.create(<SearchInput {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('useInputValue hook change inputValue when call handleInputChange', () => {
    const { result } = renderHook(() => useInputValue(() => {}));
    act(() => {
      result.current.handleInputChange(event);
    });
    expect(result.current.inputValue).toBe(event.target.value);
  });

  it('search value is modified when user change input', () => {
    const { getByTestId } = render(<SearchInput {...props} />);
    const search = getByTestId('search-input');
    expect(search.value).toBe('');
    expect(props.onSearch).toHaveBeenCalledTimes(0);
    fireEvent.change(search, event);
    expect(props.onSearch).toHaveBeenCalledWith(event.target.value);
    expect(props.onSearch).toHaveBeenCalledTimes(1);
    expect(search.value).toBe(event.target.value);
  });
});
