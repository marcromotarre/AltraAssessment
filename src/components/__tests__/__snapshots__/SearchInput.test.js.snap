// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component tests render SearchInput Snapshot 1`] = `
<div
  className="sc-AxjAm bSRsuE true"
>
  <svg
    className="sc-AxirZ hbMWzL"
    data-testid="search-icon"
    width={21.42}
  />
  <input
    className="sc-AxiKw eaUUbu"
    data-testid="search-input"
    onChange={[Function]}
    value=""
  />
</div>
`;
